File: Python3/1022_sum_of_root_to_leaf_binary_numbers.py
# https://leetcode.cn/problems/sum-of-root-to-leaf-binary-numbers/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node: Optional[TreeNode], val: int):
            nonlocal res
            
            if node.left == None and node.right == None:
                res += val
                return
            
            if node.left:
                dfs(node.left, (val << 1) + node.left.val)
            if node.right:
                dfs(node.right, (val << 1) + node.right.val)
        
        dfs(root, root.val)

        return res

File: Cpp/1022_sum_of_root_to_leaf_binary_numbers.cpp
// https://leetcode.cn/problems/sum-of-root-to-leaf-binary-numbers/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        int res = 0;

        auto dfs = [&res](auto &&self, TreeNode *node, int val) -> void {
            if ((!node->left) && (!node->right)) {
                res += val;
            }
            if (node->left) {
                self(self, node->left, (val<<1)+node->left->val);
            }
            if (node->right) {
                self(self, node->right, (val<<1)+node->right->val);
            }
        };

        dfs(dfs, root, root->val);
        return res;
    }
};